### PR TITLE
bit vector + backoff timer + simpler implementation for faster slab reassignment

### PR DESCRIPTION
**Pull request by byrnedj**
This PR addresses the implementation details for faster slab reassignment and the suggestions made by dormando. See #524 for more details.

### CODE REVIEW

File: slabs.c [hunk 1]
```diff
@@ -1018,15 +1029,25 @@ static int slab_rebalance_move(void) {
...
-                    slabs_free(it, ntotal, slab_rebal.s_clsid);
-                    /* Swing around again later to remove it from the freelist. */
-                    slab_rebal.busy_items++;
-                    was_busy++;
+                    /* unlink and mark as done if it's not
+                     * a chunked item as they require more book-keeping) */
+                    if (!ch) {
+                        do_item_unlink(it, hv);
+                        it->it_flags = ITEM_SLABBED|ITEM_FETCHED;
+                        slab_rebal.completed[offset] = 1;
```
Comment (dormando) [inline slabs.c]: think you also need refcount = 0 here, same as MOVE_FROM_SLAB below.

File: slabs.c [hunk 2]
```diff
@@ -756,6 +756,11 @@ static int slab_rebalance_start(void) {
         slab_rebal.done = 1;
     }
 
+    //bit-vector to keep track of completed chunks
```
Comment (dormando) [inline slabs.c]: // [space] comment

File: slabs.c [hunk 3]
```diff
@@ -756,6 +756,11 @@ static int slab_rebalance_start(void) {
...
+    slab_rebal.completed = (uint8_t*)malloc(
+            sizeof(uint8_t)*s_cls->perslab);
```
Comment (dormando) [inline slabs.c]: I've been using the calloc pattern in most of the codebase now. It's less error prone than if both malloc/memset lines have to be updated in sync.

File: slabs.c [hunk 4]
```diff
@@ -1018,15 +1029,25 @@ static int slab_rebalance_move(void) {
...
-                    slab_rebal.busy_items++;
-                    was_busy++;
+                    /* unlink and mark as done if it's not
+                     * a chunked item as they require more book-keeping) */
+                    if (!ch) {
+                        do_item_unlink(it, hv);
+                        it->it_flags = ITEM_SLABBED|ITEM_FETCHED;
+                        slab_rebal.completed[offset] = 1;
+                    }
+                    else {
```
Comment (dormando) [inline slabs.c]: mc standard is cuddled else :)

File: slabs.c [hunk 5]
```diff
@@ -1189,7 +1214,10 @@ static void *slab_rebalance_thread(void *arg) {
         } else if (was_busy) {
             /* Stuck waiting for some items to unlock, so slow down a bit
              * to give them a chance to free up */
-            usleep(1000);
+            usleep(backoff_timer);
```
Comment (dormando) [inline slabs.c]: curious; did you try an additive vs multiplicative backoff in your tests?

**Comment by dormando** [comment]
Thanks! Great timing, I was just about to sit down and give this a whack myself.

How did this look in your tests vs your previous patch?

File: slabs.c (hunk 5 above)
Comment (byrnedj) [reply slabs.c]: I did, but got more busy items.

**Comment by byrnedj** [comment]
It reassigns slabs at nearly the same speed. Which for all practical purposes is fast enough on our end (a few hundred GET requests for a few hundred slabs). Ran mc-crusher overnight, no issues so far.

**Comment by dormando** [comment]
hmm I really want to include this in the next release but I haven't spent much time with it. It doesn't seem like anyone else has looked at it or tried it on a shadow/production workload either.

This might miss the next release, but they're coming pretty frequently right now.

File: slabs.c [hunk 6]
```diff
@@ -1018,15 +1028,25 @@ static int slab_rebalance_move(void) {
...
+                    /* unlink and mark as done if it's not
+                     * a chunked item as they require more book-keeping) */
+                    if (!ch) {
+                        do_item_unlink(it, hv);
+                        it->it_flags = ITEM_SLABBED|ITEM_FETCHED;
+                        it->refcount = 0;
+                        slab_rebal.completed[offset] = 1;
+                    } else {
+                        ntotal = ITEM_ntotal(it);
+                        STORAGE_delete(storage, it);
```
Comment (dormando) [inline slabs.c]: Always need STORAGE_delete(), can probably move this above the if?

**Comment by byrnedj** [comment]
Sounds good, let me know if there are any other issues.

**Comment by dormando** [comment]
Fixed a couple minor nits and wrote a real commit message. It's now merged for the next release.

Think the 'next' branch will end up baking for a few days or a week, just doing a merge-down while I have a chance.

**Comment by dormando** [comment]
Out in 1.5.20

**Pull request closed without merging**